packages/full-stack-tests/src/hierarchy-builder/HierarchyLevelFiltering.test.ts

```diff
@@ -0,0 +1,371 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Element, InformationPartitionElement, Subject } from "@itwin/core-backend";
+import { IModel } from "@itwin/core-common";
+import { IHierarchyLevelDefinitionsFactory, NodeSelectQueryFactory } from "@itwin/presentation-hierarchy-builder";
+import { buildIModel, insertExternalSourceAspect, insertPhysicalPartition, insertSubject } from "../IModelUtils";
+import { initialize, terminate } from "../IntegrationTests";
+import { NodeValidators, validateHierarchyLevel } from "./HierarchyValidation";
+import { createMetadataProvider, createProvider } from "./Utils";
+
+describe("Stateless hierarchy builder", () => {
+  describe("Hierarchy filtering", () => {
+    let elementClassName: string;
+    let subjectClassName: string;
+    let partitionClassName: string;
+
+    before(async () => {
+      await initialize();
+      elementClassName = Element.classFullName.replace(":", ".");
+      subjectClassName = Subject.classFullName.replace(":", ".");
+      partitionClassName = InformationPartitionElement.classFullName.replace(":", ".");
+    });
+
+    after(async () => {
+      await terminate();
+    });
+
+    it("filters root hierarchy level", async function () {
+      const { imodel, ...keys } = await buildIModel(this, async (builder) => {
+        const rootSubject = { className: subjectClassName, id: IModel.rootSubjectId };
+        const childSubject1 = insertSubject({ builder, codeValue: "test subject 1", parentId: rootSubject.id });
+        const childSubject2 = insertSubject({ builder, codeValue: "test subject 2", parentId: rootSubject.id });
+        return { rootSubject, childSubject1, childSubject2 };
+      });
+
+      const selectQueryFactory = new NodeSelectQueryFactory(createMetadataProvider(imodel));
+      const hierarchy: IHierarchyLevelDefinitionsFactory = {
+        async defineHierarchyLevel({ instanceFilter }) {
+          const filterClauses = instanceFilter
+            ? await selectQueryFactory.createFilterClauses(instanceFilter, { fullName: subjectClassName, alias: "this" })
+            : undefined;
```

Comment:
This and 
```
FROM ${filterClauses?.from ?? subjectClassName} AS this
${filterClauses?.joins ?? ""}
```

are repeated when defining each hierarchy level. I think repetition could be reduced by doing something like this:
```
// if instance filter is undefined `createFilterClauses` returns { from: subjectClassName, joins: "" }
const filterClauses = await selectQueryFactory.createFilterClauses(instanceFilter, { fullName: subjectClassName, alias: "this" })
// ...
ecsqli: `
  ...
  FROM ${filterClauses.from} AS this
  ${filterClauses.joins}
  ...
```

Reply:
Made the following changes to `NodeSelectQueryFactory.createFilterClauses`:
- Also accept `undefined` filter definition, in which case it returns `{ from: contentClass.fullName, joins: "", where: "" }`.
- Set the clauses to empty strings rather than `undefined` when they need to be left empty - that removes the necessity for consumers to do empty string fallbacks: `filter?.joins ?? ""`.